simSPI/linear_simulator/modules.py

```diff
@@ -0,0 +1,397 @@
+        """
+        defocusU = ctf_params["defocusU"]
+        defocusV = ctf_params["defocusV"]
+        angleAstigmatism = ctf_params["defocusAngle"]
+        elliptical = (
+            defocusV * self.r2
+            + (defocusU - defocusV)
+            * self.r2
+            * torch.cos(self.angleFrequency - angleAstigmatism) ** 2
+        )
+        defocusContribution = (
+            np.pi * self.wavelength * 1e4 * elliptical * self.frequency ** 2
+        )
+        abberationContribution = (
+            -np.pi
+            / 2.0
+            * self.config.cs
+            * (self.wavelength ** 3)
+            * 1e7
+            * self.frequency ** 4
+            * self.r2 ** 2
+        )
+
+        argument = abberationContribution + defocusContribution
+
+        hFourier = (1 - self.config.amplitude_contrast ** 2) ** 0.5 * torch.sin(
+            argument
+        ) + self.config.amplitude_contrast * torch.cos(argument)
+
+        if self.config.bfactor == 0:
+            decay = (
+                np.sqrt(-np.log(self.config.valueNyquist))
+                * 2.0
+                * self.config.pixel_size
+            )
+            envelope = torch.exp(-self.frequency ** 2 * decay ** 2 * self.r2)
+        else:
+            envelope = torch.exp(
+                -self.frequency ** 2 * self.config.bfactor / 4.0 * self.r2
+            )
+
+        hFourier *= envelope
+        return hFourier
+
+    def forward(self, x_fourier, ctf_params={}):
+        """Multiply the CTF and projection in fourier domain.
+
+        If ctf_params is empty then skip multiplication.
+
+        Parameters
+        ----------
+        x_fourier: torch.Tensor
+            fourier transform of the projection (chunks,1,sidelen,sidelen)
+        ctf_params: dict of type str to {torch.Tensor}
+            contains parameters of the ctf
+
+        Returns
+        -------
+        x_fourier: torch.Tensor
+            modulated fourier transform of the projection (chunks,1,sidelen,sidelen)
+        """
+        if ctf_params:
+            hFourier = self.get_ctf(ctf_params)
+
+            x_fourier = x_fourier * hFourier
+
+        return x_fourier
+
+
+"""Module containing method to shift an image in Fourier domain."""
+
+
+class Shift(torch.nn.Module):
+    """class containing method to shift an image in Fourier domain.
+
+    Parameters
+    ----------
+    config: class
+        Class containing parameters of the shift
+
+    """
+
+    def __init__(self, config):
+        super(Shift, self).__init__()
+
+        self.config = config
+        self.frequency = 1.0 / (self.config.sidelen * self.config.pixel_size)
+
+        n2 = float(self.config.sidelen // 2)
+        ax = torch.arange(-n2, n2 + self.config.sidelen % 2)
+        ax = torch.flip(ax, dims=[0])
+        mx, my = torch.meshgrid(ax, ax)
+
+        # shape SizexSize
+        self.register_buffer("mx", mx.clone())
+        self.register_buffer("my", my.clone())
```

Comment:
what does this register_buffer do?

Reply:
Julien told me I could think of it as a way to send stuff to the GPU - he gave more details but I can only remember that part.
For more info, maybe check out: https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module.register_buffer